Beef up documentation on DispatchKey.h

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33305 Remove special case codegen for tril_indices/triu_indices.
* #33205 Bring up new-style registration API as wrapper around old-style
* #33097 Stop generating out full function type for registration, use decltype or infer it
* #33093 Delete unnecessary aliasAnalysis specification from operator registrations.
* **#33011 Beef up documentation on DispatchKey.h**

I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19796584](https://our.internmc.facebook.com/intern/diff/D19796584)

ezyang committed:
Beef up documentation on DispatchKey.h

I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

izdeby commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
   CustomRNGKeyId,
 
+  // Here are backends which specify more specialized operators
+  // based on the layout of the tensor.  Note that the sparse backends
```
ultra nit: a few occasions of double spacing.

ezyang replied:
It's the old type writer style where you double space after a period :>

izdeby commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
+  // layer which happens on top of all backends.  It inspects the autograd
+  // metadata of all inputs, determines what autograd metadata should be
+  // constructed by the output, and otherwise defers to the backend to
+  // actually do the numeric computation.
```
Should this comment be after the `VariableTensorId`? Or, at least, there should be some info about `VariableTensorId` it self

ezyang replied:
Added

```
VariableTensorId contains
    // the bulk of this logic.
```

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
+  // how to implement operations on some device.
+  CPUTensorId,    // registered at build/aten/src/ATen/CPUType.cpp
+  CUDATensorId,   // registered at build/aten/src/ATen/CUDAType.cpp
+  HIPTensorId,    // TODO: I think this is not actually used, due to Note [Masquerading as CUDA]
```
replaced with NB

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
+  //
+  // The set of backends is not really ordered (except in the case of
+  // sparse), but as our model of dispatching requires every dispatch key
+  // to participate in a total order, we have an unused ordering here.)
```
```
    // Due to the nature of the enum, these backends are specified in
    // an ordered way, but for most backends this order is not semantically
    // meaningful (e.g., it's valid to reorder these backends without changing
    // semantics).  The only situation when backend ordering is meaningful
    // is when the backend participates in multiple dispatch with another
    // backend; e.g., CPUTensorId and SparseCPUTensorId (sparse must have
    // higher priority).
```

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

gchanan commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ BACKENDS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+  // A "backend" is colloquially used to refer to handlers for dispatch
+  // which actually implement the numerics of an operation in question.
+  // Alternately, it is placed in the "back" of the dispatch hierarchy,
```
This sentence seems pretty questionable -- it wouldn't surprise me at all if we have some things running after the "backend" in the future (for tracking purposes on the output) and I don't think that's what people mean at all when they say "backend" (they just mean, it's the thing that provides the implementation for the frontend).

gchanan commented on c10/core/DispatchKey.h:
```diff
@@ -19,47 +27,90 @@ enum class DispatchKey : uint8_t {
+  CPUTensorId,    // registered at build/aten/src/ATen/CPUType.cpp
+  CUDATensorId,   // registered at build/aten/src/ATen/CUDAType.cpp
+  HIPTensorId,    // NB: I think this is not actually used, due to Note [Masquerading as CUDA]
+  MSNPUTensorId,  // tested at test/cpp_extensions/msnpu_extension.cpp
```
can we say "unused?"  I don't know if defunct refers to the key or the product.

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19796584](https://our.internmc.facebook.com/intern/diff/D19796584)

[ghstack-poisoned]

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19796584](https://our.internmc.facebook.com/intern/diff/D19796584)

[ghstack-poisoned]

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19796584](https://our.internmc.facebook.com/intern/diff/D19796584)

[ghstack-poisoned]

ezyang committed:
Update on "Beef up documentation on DispatchKey.h"


I also reordered some of the keys in non-semantic ways to make the
organizational grouping mroe clear.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19796584](https://our.internmc.facebook.com/intern/diff/D19796584)

[ghstack-poisoned]

smessmer commented on c10/core/DispatchKey.h:
```diff
@@ -19,48 +27,94 @@ enum class DispatchKey : uint8_t {
+  // semantics).  The only situation when backend ordering is meaningful
+  // is when the backend participates in multiple dispatch with another
+  // backend; e.g., CPUTensorId and SparseCPUTensorId (sparse must have
+  // higher priority).
```
I would mention that the backends further down in the list have a higher priority because I had originally assumed it otherwise, i.e. I had assumed that what's first in the list gets called first.

ezyang replied:
OK, I'll try to shuffle this clarification somewhere else

smessmer commented on c10/core/DispatchKey.h:
```diff
@@ -19,48 +27,94 @@ enum class DispatchKey : uint8_t {
+
+  // Here are backends which you think of as traditionally specifying
+  // how to implement operations on some device.
+  CPUTensorId,    // registered at build/aten/src/ATen/CPUType.cpp
```
not sure I'd mention the `Type.cpp` files. They're supposed to die and then we have this lingering comment here. Also, the ideal final state would be that CPU or CUDA tensors don't have one central file they're defined in anymore.

ezyang replied:
It's mostly to help people trying to navigate the code right now. We can update it when things change.

smessmer reviewed (approved):
Thanks, this is much needed documentation.
